src/data/post_number.rs

```diff
@@ -0,0 +1,62 @@
+use serde::{Deserialize, Serialize};
+use std::convert::TryFrom;
+
+#[derive(Serialize, Deserialize, Clone, Copy, PartialEq, Eq, Debug)]
+#[serde(try_from = "u32", into = "u32")]
+pub struct PostNumber {
+    value: u32,
+}
+
+const MIN_POST_NUMBER: u32 = 100_000_000;
+const MAX_POST_NUMBER: u32 = 999_999_999;
+
+
+impl TryFrom<u32> for PostNumber {
+    type Error = String;
+
+    fn try_from(value: u32) -> Result<Self, Self::Error> {
+        // todo: find correct constraints and enforce in here
+        if value < MIN_POST_NUMBER {
+            return Err(format!(
+                "PostNumber to small: must have exactly 9 digits {}",
+                value
+            ));
+        } else if value > MAX_POST_NUMBER {
+            return Err(format!(
+                "PostNumber to large: must have exactly 9 digits {}",
+                value
+            ));
+        } else {
+            Ok(PostNumber { value })
+        }
+    }
+}
+
+impl Into<u32> for PostNumber {
+    fn into(self) -> u32 {
+        self.value
+    }
+}
+
+impl PostNumber {
+    pub fn to_string(&self) -> String {
+        let post_number_as_string = format!("{}", self.value);
+        return format!("{} {} {}", &post_number_as_string[..3], 
+                                   &post_number_as_string[3..6], 
+                                   &post_number_as_string[6..9]);
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    
+    #[test]
+    fn should_correctly_cast_to_string() {
+        let post_number = PostNumber{ value: 123456789};
```

Comment:
An `PostNumber` instance should never be created explicitly but only via the `try_from` conversion. Right now, the `value` is not checked whether it is a valid post number.

So this line should be:

```rust
let post_number = PostNumber::try_from(123456789).unwrap();
```

I think `PostNumber{ ...}` only works in here because the test is in the same module as the struct definition. Outside of this module, the explicit construction should not work, as `value` is private: https://stackoverflow.com/questions/44865842/is-it-possible-to-make-a-private-variable-in-rust